File: kits/java/src/main/java/com/luxai/lux/Factory.java
package com.luxai.lux;

import com.fasterxml.jackson.annotation.JsonProperty;

public class Factory {

    public static final int BUILD_LIGHT = 0;
    public static final int BUILD_HEAVY = 1;
    public static final int WATER = 2;

    public static final int X = 0;
    public static final int Y = 1;

    public int[] pos;
    public int power;
    @JsonProperty("strain_id")
    public int strainId;
    @JsonProperty("team_id")
    public int teamId;

    @JsonProperty("unit_id")
    public String unitId;
    public Cargo cargo;

    public boolean canBuildHeavy(Environment environment) {
        RobotInfo robot = environment.ROBOTS.get("HEAVY");
        return this.power >= robot.POWER_COST && this.cargo.metal >= robot.METAL_COST;
    }

    public boolean canBuildLight(Environment environment) {
        RobotInfo robot = environment.ROBOTS.get("LIGHT");
        return this.power >= robot.POWER_COST && this.cargo.metal >= robot.METAL_COST;
    }

    public int waterCost(Obs obs, Environment environment) {
        int lichenCounter = 0;
        for (int[] row : obs.board.lichen_strains) {
            for (int lichenStrain : row) {
                if (lichenStrain == this.strainId)
                    lichenCounter++;
            }
        }
        return (int) (Math.ceil(lichenCounter / 10.0) * environment.LICHEN_WATERING_COST_FACTOR);
    }

    public boolean canWater(Obs obs, Environment environment) {
        return this.cargo.water >= waterCost(obs, environment);
    }

    public boolean isFactoryArea(int x, int y) {
        return Math.max(Math.abs(this.pos[X] - x), Math.abs(this.pos[Y] - y)) <= 1;
    }
}

File: kits/rl-sb3/lux/factory.py
import math
from sys import stderr
import numpy as np
from dataclasses import dataclass
from lux.cargo import UnitCargo
from lux.config import EnvConfig
@dataclass
class Factory:
    team_id: int
    unit_id: str
    strain_id: int
    power: int
    cargo: UnitCargo
    pos: np.ndarray
    # lichen tiles connected to this factory
    # lichen_tiles: np.ndarray
    env_cfg: EnvConfig

    def build_heavy_metal_cost(self, game_state):
        unit_cfg = self.env_cfg.ROBOTS["HEAVY"]
        return unit_cfg.METAL_COST
    def build_heavy_power_cost(self, game_state):
        unit_cfg = self.env_cfg.ROBOTS["HEAVY"]
        return unit_cfg.POWER_COST
    def can_build_heavy(self, game_state):
        return self.power >= self.build_heavy_power_cost(game_state) and self.cargo.metal >= self.build_heavy_metal_cost(game_state)
    def build_heavy(self):
        return 1

    def build_light_metal_cost(self, game_state):
        unit_cfg = self.env_cfg.ROBOTS["LIGHT"]
        return unit_cfg.METAL_COST
    def build_light_power_cost(self, game_state):
        unit_cfg = self.env_cfg.ROBOTS["LIGHT"]
        return unit_cfg.POWER_COST
    def can_build_light(self, game_state):
        return self.power >= self.build_light_power_cost(game_state) and self.cargo.metal >= self.build_light_metal_cost(game_state)

    def build_light(self):
        return 0

    def water_cost(self, game_state):
        """
        Water required to perform water action
        """
        owned_lichen_tiles = (game_state.board.lichen_strains == self.strain_id).sum()
        return np.ceil(owned_lichen_tiles / self.env_cfg.LICHEN_WATERING_COST_FACTOR)
    def can_water(self, game_state):
        return self.cargo.water >= self.water_cost(game_state)
    def water(self):
        return 2

    @property
    def pos_slice(self):
        return slice(self.pos[0] - 1, self.pos[0] + 2), slice(self.pos[1] - 1, self.pos[1] + 2)
